File: app/src/main/java/com/theone/music/ui/activity/PlayerActivity.java
package com.theone.music.ui.activity;//  ┏┓　　　┏┓
//┏┛┻━━━┛┻┓
//┃　　　　　　　┃
//┃　　　━　　　┃
//┃　┳┛　┗┳　┃
//┃　　　　　　　┃
//┃　　　┻　　　┃
//┃　　　　　　　┃
//┗━┓　　　┏━┛
//    ┃　　　┃                  神兽保佑
//    ┃　　　┃                  永无BUG！
//    ┃　　　┗━━━┓
//    ┃　　　　　　　┣┓
//    ┃　　　　　　　┏┛
//    ┗┓┓┏━┳┓┏┛
//      ┃┫┫　┃┫┫
//      ┗┻┛　┗┻┛

import static com.qmuiteam.qmui.arch.SwipeBackLayout.DRAG_DIRECTION_NONE;
import static com.qmuiteam.qmui.arch.SwipeBackLayout.DRAG_DIRECTION_TOP_TO_BOTTOM;
import static com.zhpan.bannerview.utils.BannerUtils.dp2px;

import android.app.Activity;
import android.content.DialogInterface;
import android.content.Intent;
import android.util.SparseArray;
import android.view.KeyEvent;
import android.view.View;
import android.widget.SeekBar;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.Observer;

import com.hjq.permissions.OnPermission;
import com.hjq.permissions.Permission;
import com.hjq.permissions.XXPermissions;
import com.hjq.toast.ToastUtils;
import com.kunminx.player.bean.dto.ChangeMusic;
import com.kunminx.player.bean.dto.PlayingMusic;
import com.qmuiteam.qmui.arch.SwipeBackLayout;
import com.qmuiteam.qmui.widget.dialog.QMUIDialog;
import com.qmuiteam.qmui.widget.dialog.QMUIDialogAction;
import com.theone.common.constant.BundleConstant;
import com.theone.common.widget.TheSelectImageView;
import com.theone.music.BR;
import com.theone.music.R;
import com.theone.music.data.model.CollectionEvent;
import com.theone.music.data.model.Music;
import com.theone.music.data.model.TestAlbum;
import com.theone.music.data.model.User;
import com.theone.music.databinding.PageMusicPlayerBinding;
import com.theone.music.player.PlayerManager;
import com.theone.music.service.DownloadService;
import com.theone.music.viewmodel.EventViewModel;
import com.theone.music.viewmodel.MusicInfoViewModel;
import com.theone.mvvm.base.BaseApplication;
import com.theone.mvvm.core.app.ext.LoaderExtKt;
import com.theone.mvvm.core.base.activity.BaseCoreActivity;
import com.theone.mvvm.ext.qmui.QMUIDialogExtKt;
import com.theone.mvvm.ext.qmui.QMUITipsDialogExtKt;

import java.util.List;

/**
 * @author The one
 * @date 2022-04-11 15:04
 * @describe TODO
 * @email 625805189@qq.com
 * @remark
 */
public class PlayerActivity extends BaseCoreActivity<MusicInfoViewModel, PageMusicPlayerBinding> {

    public static void startPlay(Activity activity, Music music) {
        Intent intent = new Intent(activity, PlayerActivity.class);
        intent.putExtra(BundleConstant.DATA, music);
        activity.startActivity(intent);
        activity.overridePendingTransition(R.anim.anim_in,0);
    }

    private EventViewModel mEvent;
    private Music mMusic;

    private boolean isTrackingTouch = false;

    protected EventViewModel getEventVm() {
        if (null == mEvent) {
            mEvent = ((BaseApplication) getApplication()).getAppViewModelProvider().get(EventViewModel.class);
        }
        return mEvent;
    }


    @Override
    public void initView(@NonNull View view) {

    }

    @Override
    public void initData() {
        // initData在createObserver之后
        mMusic = getIntent().getParcelableExtra(BundleConstant.DATA);
        if (null == mMusic) {
            return;
        }
        User user = getEventVm().getUserInfoLiveData().getValue();

        getViewModel().setLink(mMusic.shareUrl);

        PlayerManager playerManager = PlayerManager.getInstance();
        TestAlbum.TestMusic playingMusic = playerManager.getCurrentPlayingMusic();
        // TODO Step1 和当前播放的是同一个，直接拿当前的播放的数据显示
        if (null != playingMusic && playingMusic.getShareUrl().equals(mMusic.shareUrl)) {
            getViewModel().isSetSuccess().set(true);
            mMusic.pic = playingMusic.getCoverImg();
            getViewModel().setMusicInfo(mMusic, user);
            return;
        }
        // 不是同一个，就暂停上一个
        if (playerManager.isPlaying()) {
            playerManager.pauseAudio();
        }
        // 重置所有信息
        getViewModel().reset();
        String musicUrl = mMusic.getMusicUrl();
        // TODO Step2 是否有音频地址，有说明是收藏过来的
        if (!musicUrl.isEmpty()) {
            // 直接设置数据
            String cacheUrl = playerManager.getCacheUrl(musicUrl);
            getViewModel().setMediaSource(mMusic, user, !cacheUrl.isEmpty());
            return;
        }
        // TODO Step3 查询DB
        Music dbMusic = getViewModel().requestDbMusic();
        if (null != dbMusic) {
            String cacheUrl = playerManager.getCacheUrl(dbMusic.getMusicUrl());
            getViewModel().setMediaSource(dbMusic, user, !cacheUrl.isEmpty());
        }
        // TODO Step4 请求网络获取数据
        onPageReLoad();
    }

    @Override
    public void createObserver() {
        getViewModel().getResponseLiveData().observe(this, new Observer<Music>() {
            @Override
            public void onChanged(Music music) {
                LoaderExtKt.showSuccessPage(PlayerActivity.this);
                if (getViewModel().isReload()) {
                    getViewModel().setReload(false);
                    getEventVm().dispatchReloadMusic(music);
                }
                getViewModel().setMediaSource(music, getEventVm().getUserInfoLiveData().getValue(), true);
            }
        });

        getViewModel().getErrorLiveData().observe(this, new Observer<String>() {
            @Override
            public void onChanged(String s) {
                LoaderExtKt.showErrorPage(PlayerActivity.this, s, R.drawable.status_loading_view_loading_fail,
                        view -> {
                            PlayerActivity.this.onPageReLoad();
                            return null;
                        });
            }
        });

        PlayerManager playerManager = PlayerManager.getInstance();

        // 暂停事件
        playerManager.getPauseEvent().observe(this, new Observer<Boolean>() {
            @Override
            public void onChanged(Boolean pause) {
                getViewModel().isPlaying().set(!pause);
            }
        });

        // 当前播放的音乐
        playerManager.getPlayingMusicEvent().observe(this, new Observer<PlayingMusic>() {
            @Override
            public void onChanged(PlayingMusic playingMusic) {
                // 拖动进度条时不再进行设值
                if (isTrackingTouch) {
                    return;
                }
                getViewModel().setMediaCourse(playingMusic);
            }
        });

        // 更改播放音乐
        playerManager.getChangeMusicEvent().observe(this, new Observer<ChangeMusic>() {
            @Override
            public void onChanged(ChangeMusic changeMusic) {
                mEvent.dispatchPlayMusic(getCurrentMusic());
            }
        });

        // 播放错误事件
        playerManager.getPlayErrorEvent().observe(this, new Observer<String>() {
            @Override
            public void onChanged(String msg) {
                QMUITipsDialogExtKt.showFailTipsDialog(PlayerActivity.this, msg, 1000, null);
            }
        });

    }

    @Override
    public void onPageReLoad() {
        LoaderExtKt.showLoadingPage(this, "加载中");
        getViewModel().requestServer();
    }

    private Music getCurrentMusic() {
        Music cur = getViewModel().getResponseLiveData().getValue();
        if (null == cur) {
            TestAlbum.TestMusic playingMusic = PlayerManager.getInstance().getCurrentPlayingMusic();
            cur = new Music(playingMusic);
        }
        return cur;
    }

    @Override
    protected int getDragDirection(@NonNull SwipeBackLayout swipeBackLayout, @NonNull SwipeBackLayout.ViewMoveAction viewMoveAction, float downX, float downY, float dx, float dy, float slopTouch) {
        return (downY < dp2px(500) && dy >= slopTouch) ? DRAG_DIRECTION_TOP_TO_BOTTOM : DRAG_DIRECTION_NONE;
    }

    @Nullable
    @Override
    public View loaderRegisterView() {
        return getRootView();
    }

    @Nullable
    @Override
    public Object getBindingClick() {
        return new ClickProxy();
    }

    @Override
    public void applyBindingParams(@NonNull SparseArray<Object> params) {
        params.append(BR.listener, new ListenerProxy());
    }

    public class ListenerProxy implements TheSelectImageView.OnSelectChangedListener, SeekBar.OnSeekBarChangeListener {

        @Override
        public void onSelectChanged(boolean select) {
            // 收藏、添加到喜欢、取消
            // 当前是否有用户登录
            User user = getEventVm().getUserInfoLiveData().getValue();
            // 获取当前播放的数据
            Music curMusic = getCurrentMusic();
            if (null != user && null != curMusic) {
                // 一个收藏的事件
                CollectionEvent event = new CollectionEvent(select, curMusic);
                // 数据库更新收藏状态
                getViewModel().toggleCollection(user.id, event);
                // 分发一个收藏的状态
                mEvent.dispatchCollectionEvent(event);
            }
        }

        @Override
        public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
            if (fromUser) {
                getViewModel().getNowTime().set(PlayerManager.getInstance().getTrackTime(progress));
            }
        }

        @Override
        public void onStartTrackingTouch(SeekBar seekBar) {
            isTrackingTouch = true;
        }

        @Override
        public void onStopTrackingTouch(SeekBar seekBar) {
            isTrackingTouch = false;
            PlayerManager.getInstance().setSeek(seekBar.getProgress());
        }
    }

    public class ClickProxy {

        public void togglePlayPause() {
            PlayerManager.getInstance().togglePlay();
        }

        public void download() {
            requestPermission();
        }

    }

    /**
     * 请求外部存储权限
     */
    private void requestPermission() {
        XXPermissions.with(this)
                .permission(Permission.MANAGE_EXTERNAL_STORAGE)
                .constantRequest()
                .request(new OnPermission() {
                    @Override
                    public void hasPermission(List<String> granted, boolean all) {
                        startDownload();
                    }

                    @Override
                    public void noPermission(List<String> denied, boolean quick) {
                        if (quick) {
                            showPermissionFailTips(denied);
                        }
                    }
                });
    }

    /**
     * 开始下载
     */
    private void startDownload() {
        Music music = getCurrentMusic();
        ToastUtils.show("开始下载");
        DownloadService.start(this, music);
    }

    /**
     * 权限申请失败提示弹窗
     *
     * @param denied
     */
    private void showPermissionFailTips(List<String> denied) {
        QMUIDialog dialog = QMUIDialogExtKt.showMsgDialog(PlayerActivity.this, "提示", "权限被禁止，请在设置里打开权限",
                "取消", "开启权限", new QMUIDialogAction.ActionListener() {
                    @Override
                    public void onClick(QMUIDialog dialog, int index) {
                        dialog.dismiss();
                        if (index > 0) {
                            XXPermissions.startPermissionActivity(PlayerActivity.this, denied);
                        }
                    }
                }, QMUIDialogAction.ACTION_PROP_NEGATIVE);

        dialog.setCanceledOnTouchOutside(false);
        dialog.setOnKeyListener(new DialogInterface.OnKeyListener() {
            @Override
            public boolean onKey(DialogInterface dialog, int keyCode, KeyEvent event) {
                return keyCode == KeyEvent.KEYCODE_BACK && event.getRepeatCount() == 0;
            }
        });
    }

    @Override
    public void finish() {
        super.finish();
        overridePendingTransition(0,R.anim.anim_out);
    }

}

File: app/src/main/java/com/theone/music/ui/activity/PlayerActivity.kt
package com.theone.music.ui.activity

import android.app.Activity
import android.util.Log
import android.util.SparseArray
import android.view.View
import android.widget.SeekBar
import androidx.lifecycle.lifecycleScope
import com.hjq.permissions.OnPermission
import com.hjq.permissions.Permission
import com.hjq.permissions.XXPermissions
import com.hjq.toast.ToastUtils
import com.qmuiteam.qmui.arch.SwipeBackLayout
import com.qmuiteam.qmui.arch.SwipeBackLayout.DRAG_DIRECTION_NONE
import com.qmuiteam.qmui.arch.SwipeBackLayout.DRAG_DIRECTION_TOP_TO_BOTTOM
import com.qmuiteam.qmui.widget.dialog.QMUIDialogAction
import com.theone.common.callback.OnKeyBackClickListener
import com.theone.common.constant.BundleConstant
import com.theone.common.ext.*
import com.theone.music.BR
import com.theone.music.R
import com.theone.music.app.ext.toMusic
import com.theone.music.app.util.CacheUtil
import com.theone.music.data.model.CollectionEvent
import com.theone.music.data.model.Music
import com.theone.music.data.repository.DataRepository
import com.theone.music.databinding.PageMusicPlayerBinding
import com.theone.music.player.PlayerManager
import com.theone.music.service.startMusicDownloadService
import com.theone.music.ui.view.TheSelectImageView
import com.theone.music.viewmodel.EventViewModel
import com.theone.music.viewmodel.MusicInfoViewModel
import com.theone.mvvm.core.app.ext.showErrorPage
import com.theone.mvvm.core.app.ext.showLoadingPage
import com.theone.mvvm.core.app.ext.showSuccessPage
import com.theone.mvvm.core.base.activity.BaseCoreActivity
import com.theone.mvvm.ext.addParams
import com.theone.mvvm.ext.getAppViewModel
import com.theone.mvvm.ext.qmui.showFailTipsDialog
import com.theone.mvvm.ext.qmui.showMsgDialog
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

// ┏┓　  ┏┓
//┏┛┻━━━┛┻┓
//┃　　　　　　　┃
//┃　　　━　　　┃
//┃　┳┛　┗┳　┃
//┃　　　　　　　┃
//┃　　　┻　　　┃
//┃　　　　　　　┃
//┗━┓　　　┏━┛
//    ┃　　　┃                  神兽保佑
//    ┃　　　┃                  永无BUG！
//    ┃　　　┗━━━┓
//    ┃　　　　　　　┣┓
//    ┃　　　　　　　┏┛
//    ┗┓┓┏━┳┓┏┛
//      ┃┫┫　┃┫┫
//      ┗┻┛　┗┻┛
/**
 * @author The one
 * @date 2022-01-04 15:25
 * @describe TODO
 * @email 625805189@qq.com
 * @remark
 */
class PlayerActivity :
    BaseCoreActivity<MusicInfoViewModel, PageMusicPlayerBinding>() {

    companion object {

        fun startPlay(activity: Activity, music: Music) {
            activity.startActivity<PlayerActivity> {
                putExtra(BundleConstant.DATA, music)
            }
            activity.overridePendingTransition(R.anim.anim_in, 0)
        }

    }

    private val mEvent: EventViewModel by lazy { getAppViewModel<EventViewModel>() }

    private val mMusic: Music? by getValue(BundleConstant.DATA)

    private var isTrackingTouch: Boolean = false

    override fun loaderRegisterView(): View = getRootView()

    override fun getDragDirection(
        swipeBackLayout: SwipeBackLayout,
        viewMoveAction: SwipeBackLayout.ViewMoveAction,
        downX: Float,
        downY: Float,
        dx: Float,
        dy: Float,
        slopTouch: Float
    ): Int {
        return if (downY < dp2px(500) && dy >= slopTouch) {
            DRAG_DIRECTION_TOP_TO_BOTTOM
        } else DRAG_DIRECTION_NONE
    }

    /**
     * 获取当前播放的
     */
    private fun getCurrentMusic(): Music {
        return getViewModel().getResponseLiveData().value
            ?: PlayerManager.getInstance().currentPlayingMusic.toMusic()
    }

    override fun initView(root: View) {
        // 首先就是拿到传递过来的数据
        (mMusic ?: getCurrentMusic()).let { music ->
            // 赋值请求地址 https://hifini.com/thread-35837.htm
            getViewModel().link = music.shareUrl
            with(PlayerManager.getInstance()) {
                currentPlayingMusic?.run {
                    // 和当前播放的是同一个，直接拿当前的播放的数据显示
                    if (shareUrl == music.shareUrl) {
                        getViewModel().isSetSuccess.set(true)
                        getCurrentMusic().setMusicInfo()
                        return
                    }
                    // 不是同一个，就暂停上一个?
                    if (isPlaying) {
                        pauseAudio()
                    }
                }
                // 重置所有信息
                getViewModel().reset()
                // 是否有音频地址，有说明是收藏过来的
                if (music.getMusicUrl().isNotEmpty()) {
                    //直接设置数据
                    setMediaSource(music)
                    return
                }
                // 查询DB
                getViewModel().requestDbMusic()?.let {
                    val cacheUrl = getCacheUrl(it.getMusicUrl())
                    setMediaSource(it, cacheUrl.isNotEmpty())
                    return
                }
            }
            getRootView().post {
                onPageReLoad()
            }
        }
    }

    override fun createObserver() {
        // 获取播放信息成功
        getViewModel().getResponseLiveData().observe(this) {
            // 显示成功界面
            showSuccessPage()
            // 重新得到数据后，要刷新收藏里的数据
            if (getViewModel().isReload) {
                getViewModel().isReload = false
                // 从新获取新的播放地址，但是收藏界面的数据并没有更新
                // 避免下一次点击时依然是旧的数据，会造成重复数据加载
                mEvent.dispatchReloadMusic(it)
            }
            setMediaSource(it, true)
        }
        getViewModel().getErrorLiveData().observe(this) {
            showErrorPage(it) {
                onPageReLoad()
            }
        }

        with(PlayerManager.getInstance()) {

            // 暂停事件
            pauseEvent.observe(this@PlayerActivity) {
                getViewModel().isPlaying.set(!it)
            }

            // 当前播放的音乐
            playingMusicEvent.observe(this@PlayerActivity) {
                // 拖动进度条时不再进行设值
                if (isTrackingTouch) {
                    return@observe
                }
                getViewModel().run {
                    // 只有在设置了音乐数据后才能设置播放信息，避免被上一首的播放信息污染
                    if (!isSetSuccess.get()) {
                        return@observe
                    }
                    max.set(it.duration)
                    progress.set(it.playerPosition)
                    nowTime.set(it.nowTime)
                    allTime.set(it.allTime)
                }
            }

            // 更改播放音乐
            changeMusicEvent.observe(this@PlayerActivity) {
                mEvent.dispatchPlayMusic(getCurrentMusic())
            }

            // 播放错误
            playErrorEvent.observe(this@PlayerActivity) {
                showFailTipsDialog(it)
            }

        }
    }

    private fun Music.setMusicInfo() {
        getViewModel().let {
            it.isSuccess.set(true)
            it.isCollectionEnable.set(CacheUtil.isLogin())
            it.name.set(title)
            it.author.set(author)
            it.cover.set(pic)
            mEvent.getUserInfoLiveData().value?.let { user ->
                it.requestCollection(user.id, shareUrl)
            }
        }
    }

    /**
     * 设置播放资源
     * @param data Music
     * @param newData Boolean 是否为请求来的新数据
     */
    private fun setMediaSource(data: Music, newData: Boolean = false) {
        data.setMusicInfo()
        lifecycleScope.launch {
            withContext(Dispatchers.IO) {
                val cacheUrl = PlayerManager.getInstance().getCacheUrl(data.getMusicUrl())
                // 缓存有就直接播放
                if (cacheUrl.isNullOrEmpty()) {
                    loadAlbum(data)
                } else {
                    // 不是新数据才检查地址是否可行
                    if (!newData && getViewModel().checkUrl(data.getMusicUrl())) {
                        getViewModel().isReload = true
                        getViewModel().requestServer()
                    } else {
                        loadAlbum(data)
                    }
                }
            }
        }
    }

    private fun loadAlbum(data: Music) {
        PlayerManager.getInstance().loadAlbum(data)
        getViewModel().run {
            isSetSuccess.set(true)
            // 更新音乐信息 最后播放时间
            updateMusicLastPlayDate()
        }
    }

    override fun onPageReLoad() {
        showLoadingPage()
        getViewModel().requestServer()
    }

    override fun SparseArray<Any>.applyBindingParams() {
        addParams(BR.listener, ListenerProxy())
    }

    override fun getBindingClick(): Any = ClickProxy()


    inner class ListenerProxy : TheSelectImageView.OnSelectChangedListener,
        SeekBar.OnSeekBarChangeListener {

        override fun onSelectChanged(isSelected: Boolean) {
            // 收藏、添加到喜欢、取消
            // 当前是否有用户登录
            mEvent.getUserInfoLiveData().value?.let { user ->
                // 获取当前播放的数据
                getCurrentMusic().let {
                    // 一个收藏的事件
                    CollectionEvent(isSelected, it).let { event ->
                        // 数据库更新收藏状态
                        getViewModel().toggleCollection(user.id, event)
                        // 分发一个收藏的状态
                        mEvent.dispatchCollectionEvent(event)
                    }
                }
            }
        }

        override fun onProgressChanged(seekBar: SeekBar?, progress: Int, fromUser: Boolean) {
            if (fromUser) {
                getViewModel().nowTime.set(PlayerManager.getInstance().getTrackTime(progress))
            }
        }

        override fun onStartTrackingTouch(seekBar: SeekBar?) {
            isTrackingTouch = true
        }

        override fun onStopTrackingTouch(seekBar: SeekBar?) {
            isTrackingTouch = false
            seekBar?.let {
                PlayerManager.getInstance().setSeek(it.progress)
            }
        }

    }


    private fun showFailDialog(
        tips: String,
        actionString: String,
        callback: (() -> Unit?)? = null
    ) {
        showMsgDialog(
            "提示",
            tips,
            rightAction = actionString,
            listener = QMUIDialogAction.ActionListener { dialog, index ->
                dialog.dismiss()
                callback?.invoke()
            },
            prop = QMUIDialogAction.ACTION_PROP_NEGATIVE
        ).run {
            setCanceledOnTouchOutside(false)
            setOnKeyListener(OnKeyBackClickListener())
        }
    }

    /**
     * 请求权限
     *  网络请求的缓存，是放在APP/data目录下的，这个地址是不需要权限申请的
     *
     *
     *
     */
    private fun requestPermission() {
        XXPermissions.with(this)
            .permission(Permission.MANAGE_EXTERNAL_STORAGE)
            .constantRequest()
            .request(object : OnPermission {

                override fun hasPermission(granted: MutableList<String>?, all: Boolean) {
                    startDownload()
                }

                override fun noPermission(denied: MutableList<String>?, quick: Boolean) {
                    if (quick) {
                        showFailDialog(
                            "权限被禁止，请在设置里打开权限",
                            "打开权限"
                        ) {
                            XXPermissions.startPermissionActivity(this@PlayerActivity, denied)
                        }
                    }
                }
            })
    }

    /**
     * 下载：
     *
     *  缺点：
     *  由于时间原因，下载没有做是否已经下载判断
     */
    private fun startDownload() {
        getCurrentMusic().let {
            ToastUtils.show("开始下载")
            startMusicDownloadService(it)
        }
    }

    inner class ClickProxy {

        fun togglePlayPause() {
            PlayerManager.getInstance().togglePlay()
        }

        fun download() {
            requestPermission()
        }

    }

}
